### DISCUSSION TITLE
Focus windows on middle & right button press

### DISCUSSION THREAD

**Discussion by 5ila5** [qtile/qtile#3415] [Q&A]
Is there a way to focus clicked windows on mouse wheel press or right click? 
when ever I want to open a new tab in my browser on my secondary monitor I middle click in the tab part of my browser to open a new tab but when I start typing there is an other window focused.

**Comment by SweepingBishops** [comment]
There is a list in the config named mouse. You can change it how you want there.
From mine
```
mouse = [
    Drag([mod], "Button1", move_snap_window(snap_dist=40),
         start=lazy.window.get_position()),
    Drag([mod], "Button3", lazy.window.set_size_floating(),
         start=lazy.window.get_size()),
    Click([mod], "Button2", lazy.window.bring_to_front())
]
```

Note that Button1 is left click, Button2 is middle and Button3 is right.

**Reply by 5ila5** [reply]
That's not what I want to achieve for one there is no lazy function to focus the currently hovered window (as far as I can tell). And the Click mouse event swallows my click what I want to achieve is:

current focus a terminal on monitor 2. When I click (Button2) in the tab bar of my browser to open a new tab the browser window should be focused AND a new tab should open. 

But currently the terminal window is still focused and a new browser tab opens (until I Button1 click the browser it of course)

**Reply by SweepingBishops** [reply]
For me, just hovering the mouse over an open window brings focus to it. So unless you clicked on the brower and then moved the mouse over the terminal, the focus should be on the browser. If that is not the case, I don't know what's wrong :(

**Reply by elParaguayo** [reply]
That sounds like you've got `follow_mouse_focus = True` in your config.

@5ila5 does the browser focus when you right-click on it?

**Reply by SweepingBishops** [reply]
> That sounds like you've got `follow_mouse_focus = True` in your config.

Oh yeah, completely forgot about that.

**Reply by 5ila5** [reply]
@elParaguayo I Know this as well but then the focus always follows my mouse but I only want to change focus on Button1/Button2/Button3 click

and No the focus also does not change on right-click

**Reply by SweepingBishops** [reply]
> current focus a terminal on monitor 2. When I click (Button2) in the tab bar of my browser to open a new tab the browser window should be focused AND a new tab should open.

Now that I think about it, wouldn't you have to press the middle button twice anyway? Once to change focus and then to open a tab?

**Reply by 5ila5** [reply]
When I left click a button in a not focused widow the button gets pressed and the focus switches to this window. And I want the same thing for Button2 and Button3

**Reply by elParaguayo** [reply]
Interesting. The code that focuses windows/screens on a click event doesn't filter out the button type. I'd need to take a closer look.

**Reply by 5ila5** [reply]
First click into Firefox is left click. The second click into Firefox is middle mouse button.

https://user-images.githubusercontent.com/38183212/157881623-1389674d-2e75-4941-a330-a68746f258f0.mp4

**Reply by SweepingBishops** [reply]
Yeah, I tested it. Works the same. A bit weird how this is only the case for middle click and not left or right.

**Reply by 5ila5** [reply]
I think the window only grabs only Button1

https://github.com/qtile/qtile/blob/68fb7da703fd6466c53f76eae5983eca35d54503/libqtile/backend/x11/window.py#L774-L795

**Reply by elParaguayo** [reply]
Yeah, was just looking at that bit. Can try adding other buttons to the method and test later.

Looks like there's an `AnyButton` (`xcffib.xproto.ButtonIndex.Any`) option in the X specification which may be what we need here.

**Reply by elParaguayo** [reply]
Just tested - can now focus window with right and middle clicks. Also works for opening tab with middle click in unfocused window.

**Reply by 5ila5** [reply]
@elParaguayo thanks works for me as well.
Will it be implemented like that? Should I create a pull request?

**Reply by elParaguayo** [reply]
I'm wondering what the best way to implement this is. Some people may not want to focus windows with any mouse button event (e.g. scrolling) but that would mean adding another config option which is something I try to avoid.

I could submit a PR to force the discussion with other members of the team!

**Reply by 5ila5** [reply]
I see. Changing focus on scrolling is annoying as well i would prefer to just grab buttons 1 to 3 I think this should annoy no one (https://github.com/qtile/qtile/pull/3459)
